oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizationSuccessHandler.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * Handles when an OAuth 2.0 Authorized Client
+ * has been successfully authorized (or re-authorized)
+ * via the authorization server.
+ *
+ * @author Phil Clay
+ * @since 5.3
+ * @see OAuth2AuthorizationSuccessEvent
+ */
+@FunctionalInterface
+public interface ReactiveOAuth2AuthorizationSuccessHandler {
+
+	/**
+	 * Called when an OAuth 2.0 Authorized Client
+	 * has been successfully authorized (or re-authorized)
+	 * via the authorization server.
+	 *
+	 * @param authorizationSuccessEvent details of the successful authorization event that occurred
+	 * @return an empty {@link Mono} that completes after this handler has finished handling the event.
+	 */
+	Mono<Void> onAuthorizationSuccess(OAuth2AuthorizationSuccessEvent authorizationSuccessEvent);
```

Comment:
As I mentioned during the design iteration, I wasn't convinced with the `*Event` naming. Event objects are used in a pub-sub interaction model, which is not the case here. The success/failure handlers are meant to be direct collaborators of the manager and therefore are coupled in the design. I just get a feel that using `*Event` in the class name and `*notify*` in internal methods will cause some confusion with the API. 

The simple option is to change the name. I can't think of a better name at the moment.

The other option is we flatten the arguments. For example:

```java
public interface ReactiveOAuth2AuthorizationSuccessHandler {

	Mono<Void> onAuthorizationSuccess(OAuth2AuthorizedClient authorizedClient, Authentication principal, Map<String, Object> attributes);

	default Mono<Void> onAuthorizationSuccess(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
		return onAuthorizationSuccess(authorizedClient, principal, Collections.emptyMap());
	}
}
```

What are your thoughts?

Reply:
I can't think of a better name for the class.  I'm fine with flattening the args as you have proposed.  I'll change it and see what it looks like.

Reply:
@jzheaux @rwinch Can you review this [comment](https://github.com/spring-projects/spring-security/pull/7756#discussion_r363483374) and suggest the approach you are in favour of? Either rename the input to the handler operations or flatten the arguments. I'm kind of divided with this one. I think I prefer the one input object as it will make it easier to add a new attribute (later on) without changing the interface.

Reply:
@jgrandja Agreed that we don't want to use `*Event` since it could easily be confused with `ApplicationEventPublisher`, `AuthenticationEventPublisher`, et al. 

Given that `AuthenticationSuccessHandler` looks like this:

```
void onAuthenticationSuccess(HttpServletRequest request,
			HttpServletResponse response, Authentication authentication)
```

And `AuthenticationFailureHandler` looks like this:

```java
void onAuthenticationFailure(HttpServletRequest request,
			HttpServletResponse response, AuthenticationException exception)
```

And that this pattern is continued in the reactive stack, I'd favor modeling these other Spring Security success/failure handlers more closely (as you already said "flatten the arguments").

Reply:
Flattened the args as requested.